File: leetcode/easy/20. Valid Parentheses/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
    const stack = [];
    const table = {
        '(': ')',
        '{': '}',
        '[': ']',
    };
    for (let i = 0; i < s.length; i += 1) {
        if (table[s[i]]) {
            stack.push(table[s[i]]);
            continue;
        }
        if (s[i] !== stack.pop()) return false;
    }
    if (stack.length > 0) return false;
    return true;
};

console.log(isValid("()"))
console.log(isValid("()["))

File: leetcode/easy/20. Valid Parentheses/solution.cc
#include <string>
#include <map>
#include <stack>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        map<char, char> map;
        map['('] = ')';
        map['{'] = '}';
        map['['] = ']';

        for (int i = 0; i < s.size(); i++) {
            if (map[s[i]]) {
                stack.push(map[s[i]]);
                continue;
            }
            else {
                if (stack.size() == 0) return false;
                if (stack.top() != s[i]) return false;
                stack.pop();
            }
        }
        return stack.size() == 0;
    }
};